llvm/lib/SYCLLowerIR/ModuleSplitter.cpp

```diff
@@ -907,305 +907,131 @@ std::string ModuleDesc::makeSymbolTable() const {
-  // function, which will be used as an entry points group identifier.
-  void registerRule(const std::function<std::string(Function *)> &Callback) {
-    Rules.emplace_back(Rule::RKind::K_Callback, Callback);
-  }
-
-  // Creates a simple rule, which adds a value of a string attribute into a
-  // resulting identifier.
-  void registerSimpleStringAttributeRule(StringRef AttrName) {
-    Rules.emplace_back(Rule::RKind::K_SimpleStringAttribute, AttrName);
-  }
 
-  // Creates a simple rule, which adds a value of a string metadata into a
-  // resulting identifier.
-  void registerSimpleStringMetadataRule(StringRef MetadataName) {
-    Rules.emplace_back(Rule::RKind::K_SimpleStringMetadata, MetadataName);
+void computeFuncCategoryFromAttribute(const Function &F, StringRef AttrName,
+                                      SmallString<256> &Result) {
+  if (F.hasFnAttribute(AttrName)) {
+    const Attribute &Attr = F.getFnAttribute(AttrName);
+    Result += Attr.getValueAsString();
   }
 
-  // Creates a simple rule, which adds one or another value to a resulting
-  // identifier based on the presence of a metadata on a function.
-  void registerSimpleFlagAttributeRule(StringRef AttrName,
-                                       StringRef IfPresentStr,
-                                       StringRef IfAbsentStr = "") {
-    Rules.emplace_back(Rule::RKind::K_FlagAttribute,
-                       Rule::FlagRuleData{AttrName, IfPresentStr, IfAbsentStr});
-  }
-
-  // Creates a simple rule, which adds one or another value to a resulting
-  // identifier based on the presence of a metadata on a function.
-  void registerSimpleFlagMetadataRule(StringRef MetadataName,
-                                      StringRef IfPresentStr,
-                                      StringRef IfAbsentStr = "") {
-    Rules.emplace_back(
-        Rule::RKind::K_FlagMetadata,
-        Rule::FlagRuleData{MetadataName, IfPresentStr, IfAbsentStr});
-  }
-
-  // Creates a rule, which adds a list of dash-separated integers converted
-  // into strings listed in a metadata to a resulting identifier.
-  void registerListOfIntegersInMetadataRule(StringRef MetadataName) {
-    Rules.emplace_back(Rule::RKind::K_IntegersListMetadata, MetadataName);
-  }
+  Result += "-";
+}
 
-  // Creates a rule, which adds a list of sorted dash-separated integers
-  // converted into strings listed in a metadata to a resulting identifier.
-  void registerListOfIntegersInMetadataSortedRule(StringRef MetadataName) {
-    Rules.emplace_back(Rule::RKind::K_SortedIntegersListMetadata, MetadataName);
+void computeFuncCategoryFromStringMetadata(const Function &F,
+                                           StringRef MetadataName,
+                                           SmallString<256> &Result) {
+  if (F.hasMetadata(MetadataName)) {
+    const auto *MDN = F.getMetadata(MetadataName);
+    for (size_t I = 0, E = MDN->getNumOperands(); I < E; ++I) {
```

Comment:
Any reason to do this instead of the following?
```suggestion
    for (size_t I = 0; I < MDN->getNumOperands(); ++I) {
```

Reply:
This is an accepted approach in the community due to https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop